### PR TITLE
RHSTOR-1554: Migrate object service dashboard tests 

### PR DESCRIPTION
**Pull request by rexagod**
Based-off-of: https://github.com/openshift/console/pull/7815

Fixes: https://issues.redhat.com/browse/RHSTOR-1554

### CODE REVIEW

**Comment by rexagod** [comment]
@bipuladh I've made the changes suggested in the review.

PTAL. Thanks!

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+describe('Tests Buckets Card', () => {
+  before(() => {
+    installStorageCluster();
+  });
+
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+    cy.wait(10 * SECOND);
```
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:36]: Confirmed this a couple of times and both `it` need a 10 second buffer time for the page to load.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+describe('Test Object Storage Efficiency Card', () => {
+  before(() => {
+    installStorageCluster();
+  });
+
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+    cy.wait(10 * SECOND);
```
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:99]: > Confirmed this a couple of times and both it need a 10 second buffer time for the page to load.

Same here.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 1 above)
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:36]: Wait for the elements that you want to test. For example if you want to test the details card then wait on the details card to load. Don't wait arbitrarily.

**Comment by rexagod** [comment]
Tests look good on local.

image

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 1 above)
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:36]: Right, that's better. Sending a patch soon.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      .byTestID('buckets-status-title')
+      .invoke('text')
+      .then((text) => {
+        const count = parseInt(text.split(' ')[0], 10);
+        expect(count).toBeGreaterThanOrEqual(1);
+      });
+  });
+
+  it('Create an Object Bucket Claim and test equality', () => {
+    cy.wait(10 * SECOND)
```
Comment (vbnrh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:49]: As per best practices, arbitrary wait times are not allowed. This is one of the issues i faced in my PR, you probably need to look into this

**Review by vbnrh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 3 above)
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:49]: Thanks for pointing this out! Sending a patch soon based on [this](https://docs.cypress.io/guides/references/best-practices.html#Unnecessary-wait-for-cy-get) soon.

Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:49]: It seems the above method is only for routes. Using `{ timeout: __ }` instead.

**Comment by rexagod** [comment]
I may very well be wrong but it seems like the `/dashboards/object-service` page loads twice, which is the reason why `cy.wait` is necessary. Not sure why though.

**Comment by rexagod** [comment]
@bipuladh @vbnrh Any leads on how to fetch the current state for these three from the CLI? Would be a great help! Thanks!

image

**Comment by rexagod** [comment]
Okay I think I got it. 

image

**Comment by rexagod** [comment]
Tried this with @bipuladh's CLI suggestion, worked on the first try! :)

image

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    installStorageCluster();
+  });
+
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+  });
+
+  it('Test at least one noobaa bucket is present', () => {
+    cy.byTestID('1-obc-test')
```
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:37]: This assumes exactly `1 NooBaa Bucket` for the test, only for the time being. Will replace this with a better alternative once I figure out how to get the number of "NooBaa Bucket" via CLI.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  },
+};
+
+const installStorageCluster = () => {
+  cy.login();
+  cy.visit('/');
+  cy.install();
+};
+
+describe('Tests Buckets Card', () => {
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:26]: You don't need multiple describe blocks. Just use one describe block and separate different cards into various 'it' blocks.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  });
+
+  it('Test at least one noobaa bucket is present', () => {
+    cy.byTestID('1-obc-test')
+      .invoke('text')
+      .should('match', new RegExp('\\d+\\sNooBaa\\sBucket'));
+  });
+
+  it('Create an Object Bucket Claim and test equality', () => {
+    cy.exec(`kubectl get ObjectBucketClaims -A | wc -l`).then((outObj) => {
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:43]:
```suggestion
    cy.exec(`kubectl get ObjectBucketClaims -A | wc -l`).then(({stdout})) => {
```

**Review by bipuladh — CHANGES_REQUESTED** [review]
Need to look in-depth, for now we can start with these changes

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 5 above)
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts]: Do you mean that I should covert existing additional describe blocks into it blocks and remove the sub-it blocks that are inside those? I've pushed a patch that simply removes the additional describes but keeps their it blocks.

Let me know what changes you have in mind and I'll push them.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 4 above)
Comment (bipuladh) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:37]: Noobaa buckets are queried via Prometheus.

**Comment by rexagod** [comment]
Fixed the hardcoded "x NooBaa Bucket" in this commit (https://github.com/openshift/console/pull/8270/commits/41a827f93d807aaf95c05dba3cf89595f6838fa6).

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 4 above)
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts]: Fixed in https://github.com/openshift/console/commit/41a827f93d807aaf95c05dba3cf89595f6838fa6.

**Comment by rexagod** [comment]
- https://github.com/openshift/console/pull/8270/commits/e15769c8d28d1c5a5cd032c0678b4b2198af9e83: Moved describe blocks to it blocks, and it blocks to `cy.log`.
- https://github.com/openshift/console/pull/8270/commits/b7710aeefe6958ee43bcc0dc7f1d44febc753f8d: Moved some constants to `consts.ts`.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+const installStorageCluster = () => {
+  cy.login();
+  cy.visit('/');
+  cy.install();
+};
+
+describe('Tests Buckets, Status, Object Storage Efficiency, and Resource Providers Cards', () => {
+  before(() => {
+    installStorageCluster();
```
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:11]: Move `installStorageCluster` in here.

File: frontend/packages/console-shared/src/components/dashboard/inventory-card/InventoryItem.tsx [hunk 8]
```diff
@@ -118,7 +118,15 @@ export const InventoryItem: React.FC<InventoryItemProps> = React.memo(
           data-test={!TitleComponent ? dataTest : null}
         >
           {isLoading && !error && <div className="skeleton-inventory" />}
-          {TitleComponent ? <TitleComponent>{titleMessage}</TitleComponent> : titleMessage}
+          {TitleComponent ? (
+            <TitleComponent>
+              <div data-test={`${parseInt(titleMessage, 10).toString()}-obc-test`}>
```
Comment (rexagod) [inline frontend/packages/console-shared/src/components/dashboard/inventory-card/InventoryItem.tsx:123]: Remove.

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  });
+
+  beforeEach(() => {
+    cy.visit('/');
+    cy.byLegacyTestID('horizontal-link-Object Service').click();
+  });
+
+  it('Tests Buckets Cards', () => {
```
Comment (rexagod) [inline frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts:19]: TODO.

**Comment by rexagod** [comment]
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/8270/pull-ci-openshift-console-master-e2e-gcp-console/1372817923416002560/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/

**Review by bipuladh — APPROVED** [review]
/lgtm
/assign @rawagner

File: frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rexagod) [reply frontend/packages/ceph-storage-plugin/integration-tests-cypress/tests/object-service-dashboards.spec.ts]: I think you're viewing an older commit. I've done this in the [latest commit](https://github.com/openshift/console/pull/8270/files#diff-f5ff9081f8c3177bae430651bc5e1879b0a31216c50bbae818e41d100e46982dR5).

**Comment by rexagod** [comment]
@bipuladh Can I go ahead and squash these commits now?

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx [hunk 10]
```diff
@@ -5,28 +5,28 @@ import { DashboardCardPopupLink } from '../dashboard-card/DashboardCardLink';
...
-  <div className="co-dashboard-icon">
+const HealthItemIcon: React.FC<HealthItemIconProps> = ({ state, dataTest }) => (
+  <div data-test={dataTest} className="co-dashboard-icon">
     {(healthStateMapping[state] || healthStateMapping[HealthState.UNKNOWN]).icon}
   </div>
 );
 
 const HealthItem: React.FC<HealthItemProps> = React.memo(
-  ({ className, state, title, details, popupTitle, noIcon = false, children }) => {
+  ({ className, state, title, details, popupTitle, noIcon = false, dataTest, children }) => {
```
Comment (rawagner) [inline frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:15]: I dont see any code calling `HealthItem` with this `dataTest` prop .. ?

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx [hunk 11]
```diff
@@ -5,28 +5,28 @@ import { DashboardCardPopupLink } from '../dashboard-card/DashboardCardLink';
...
     const { t } = useTranslation();
 
     const detailMessage = details || healthStateMessage(state, t);
 
     return (
-      <div className={classNames('co-status-card__health-item', className)}>
+      <div className={classNames('co-status-card__health-item', className)} data-test={dataTest}>
```
Comment (rawagner) [inline frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:21]: cant we just generate the data-test value to be something like
```suggestion
      <div className={classNames('co-status-card__health-item', className)} data-test={`${title}-health-item`}>
```

we wouldnt need to pass the `dataTest` prop

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx (hunk 10 above)
Comment (rexagod) [reply frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx]: Right, fixed!

File: frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx (hunk 11 above)
Comment (rexagod) [reply frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx]: Fixed!

**Pull request merged**